integration_tests/sdk/aqueduct_tests/naming_test.py

```diff
@@ -33,87 +67,258 @@ def test_extract_with_default_name_collision(client, data_integration):
+    assert output2.name() == "foo artifact"
+    assert output3.name() == "foo artifact"
+    assert output4.name() == "foo artifact"
+
+    flow = publish_flow_test(
+        client, artifacts=[output1, output2, output3, output4], engine=engine, name=flow_name()
+    )
+    flow_run = flow.latest()
+    assert flow_run.artifact("foo artifact").get() == 123
+    assert flow_run.artifact("foo artifact (1)").get() == "hello"
+    assert flow_run.artifact("foo artifact (2)").get() == 123
+    assert flow_run.artifact("foo artifact (3)").get() == "hello"
+
 
+def test_operator_with_explicit_artifact_naming_collision(client, engine, flow_name):
+    @op(outputs=["output1", "output2"])
+    def foo():
+        return 123, "hello"
+
+    output1, output2 = foo()
+    output3, output4 = foo()
 
-def test_extract_with_default_artifact_name_collision(client, data_integration):
-    # The default output artifact name can also collide with an existing artifact!
-    client.create_param("sql query artifact", default=123)
     with pytest.raises(
         InvalidUserActionException,
-        match="has already been created locally. Artifact names must be unique.",
+        match="Unable to publish flow. You are attempting to publish multiple artifacts explicitly named `output1`",
     ):
-        extract(data_integration, DataObject.SENTIMENT, op_name="sql query")
+        client.publish_flow("Test", artifacts=[output1, output2, output3, output4], engine=engine)
 
+    # Trigger another collision case with the artifact.set_name() method.
+    @op
+    def bar():
+        return 555
 
-def test_function_with_name_collision(client, data_integration):
-    """Colliding functions are always overwritten."""
-    table_artifact = extract(data_integration, DataObject.SENTIMENT, op_name="sql query")
+    bar_output = bar()
+    bar_output.set_name("output1")
+    with pytest.raises(
+        InvalidUserActionException,
+        match="Unable to publish flow. You are attempting to publish multiple artifacts explicitly named `output1`",
+    ):
+        client.publish_flow("Test", artifacts=[output1, bar_output], engine=engine)
 
-    # There's not an easy way to programmatically change the function, so lets
-    # just run the same function twice and check that the latest one wins.
-    dummy_fn_artifact_old = dummy_model(table_artifact)
-    dummy_fn_artifact_new = dummy_model(table_artifact)
 
-    with pytest.raises(ArtifactNotFoundException):
-        dummy_fn_artifact_old.get()
+def test_explicit_and_implicit_artifact_name_collisions(client, engine, flow_name):
+    """Test that if such a collision occurs, the explicit name always wins."""
 
-    fn_df = dummy_fn_artifact_new.get()
-    assert list(fn_df) == [
-        "hotel_name",
-        "review_date",
-        "reviewer_nationality",
-        "review",
-        "newcol",
-    ]
-    assert fn_df.shape[0] == 100
+    @op
+    def foo():
+        return 123
+
+    @op(outputs=["foo artifact"])
```

Comment:
If I explicitly name my artifact sth with a suffix "foo (1)", does the naming resolution still work? Unlikely to happen but just curious.

Reply:
Yeah it should! See `test_bump_artifact_suffix`. Just to be clear though, are you trying to collide an artifact explicitly named `foo artifact (1)` with another one that is implicitly named `foo artifact (1)`? I can test it out but that's a three artifact situation.

Reply:
Yeah that's what I was thinking. Will a explicitly named `foo artifact (1)` and our suffix bumping logic to generate an implicit name `foo artifact (1)` work, and yes it's a 3 artifact situation.